app/javascript/components/upload/UploadWizard.js

```diff
@@ -0,0 +1,198 @@
+import ClusteringStep from './ClusteringStep'
+import ImageStep from './ImageStep'
+import CoordinateLabelStep from './CoordinateLabelStep'
+import RawCountsStep from './RawCountsStep'
+import ProcessedExpressionStep from './ProcessedExpressionStep'
+
+const STEPS = [RawCountsStep, ProcessedExpressionStep, ClusteringStep, CoordinateLabelStep, ImageStep]
+
+/** shows the upload wizard */
+export default function UploadWizard({ accession, name }) {
+  const [currentStep, setCurrentStep] = useState(STEPS[0])
+  const [serverState, setServerState] = useState(null)
+  const [formState, setFormState] = useState(null)
+
+  // go through the files and compute any relevant derived properties, notably 'isDirty'
+  if (serverState?.files && formState?.files) {
+    formState.files.forEach(file => {
+      const serverFile = serverState.files.find(sFile => sFile._id === file._id)
+      // use isMatch to confirm that specified properties are equal, but ignore properties (like isDirty)
+      // that only exist on the form state objects
+      if (!_isMatch(file, serverFile)) {
+        file.isDirty = true
+      }
+    })
+  }
+
+  /** adds an empty file, merging in the given fileProps. Does not communicate anything to the server */
+  function addNewFile(fileProps) {
+    const newFile = {
+      study_id: serverState.study._id.$oid,
+      name: '',
+      _id: _uniqueId('newFile-'), // we just need a temp id to give to form controls, the real id will come from the server
+      status: 'new',
+      description: '',
+      parse_status: 'unparsed',
+      spatial_cluster_associations: []
+    }
+    Object.assign(newFile, fileProps)
+
+    setFormState(prevFormState => {
+      const newState = _cloneDeep(prevFormState)
+      newState.files.push(newFile)
+      return newState
+    })
+  }
+
+  /** handle response from server after an upload by updating the serverState with the updated file response */
+  function handleSaveResponse(response) {
+    const updatedFile = formatFileFromServer(response)
+    // first update the serverState
+    setServerState(prevServerState => {
+      const newServerState = _cloneDeep(prevServerState)
+      const fileIndex = newServerState.files.findIndex(f => f._id === updatedFile._id)
+      newServerState.files[fileIndex] = updatedFile
+      return newServerState
+    })
+    // then update the form state
+    setFormState(prevFormState => {
+      const newFormState = _cloneDeep(prevFormState)
+      const fileIndex = newFormState.files.findIndex(f => f._id === updatedFile._id)
+      newFormState.files[fileIndex] = updatedFile
+      return newFormState
+    })
+  }
+
+  /** Updates file fields by merging in the 'updates', does not perform any validation, and
+   *  does not save to the server */
+  function updateFile(fileId, updates) {
+    setFormState(prevFormState => {
+      const newFormState = _cloneDeep(prevFormState)
+      const fileChanged = newFormState.files.find(file => file._id === fileId)
```

Comment:
For consistency with two very similar code blocks above:

```suggestion
      const fileChanged = newFormState.files.find(f => f._id === fileId)
```